File: app/src/main/java/com/billpayments/features/auth/data/SessionRepository.kt
package com.billpayments.features.auth.data

import com.billpayments.core.storage.KeyValueStore

data class Session(val userId: String, val phone: String)

interface SessionRepository {
    fun restore(): Session?
    suspend fun save(userId: String, phone: String)
    suspend fun clear()
}

class StoredSessionRepository(private val store: KeyValueStore) : SessionRepository {

    override fun restore(): Session? {
        val userId = store.read(USER_ID_KEY) ?: return null
        val phone = store.read(PHONE_KEY) ?: return null
        return Session(userId, phone)
    }

    override suspend fun save(userId: String, phone: String) {
        store.write(USER_ID_KEY, userId)
        store.write(PHONE_KEY, phone)
    }

    override suspend fun clear() {
        store.remove(USER_ID_KEY)
        store.remove(PHONE_KEY)
    }

    private companion object {
        const val USER_ID_KEY = "session.userId"
        const val PHONE_KEY = "session.phone"
    }
}

File: BillPayments/Features/Auth/Data/SessionRepository.swift
struct Session: Equatable {
    let userId: String
    let phone: String
}

protocol SessionRepository {
    func restore() -> Session?
    func save(userId: String, phone: String) async throws
    func clear() async throws
}

final class StoredSessionRepository: SessionRepository {
    private static let userIdKey = "session.userId"
    private static let phoneKey = "session.phone"

    private let store: KeyValueStore

    init(store: KeyValueStore) {
        self.store = store
    }

    func restore() -> Session? {
        guard let userId = store.read(Self.userIdKey),
              let phone = store.read(Self.phoneKey) else { return nil }
        return Session(userId: userId, phone: phone)
    }

    func save(userId: String, phone: String) async throws {
        try await store.write(Self.userIdKey, userId)
        try await store.write(Self.phoneKey, phone)
    }

    func clear() async throws {
        try await store.remove(Self.userIdKey)
        try await store.remove(Self.phoneKey)
    }
}
